### DISCUSSION TITLE
TypeError: __hash__ method should return an integer

### DISCUSSION THREAD

**Discussion by mikehagerty** [sqlalchemy/sqlalchemy#7098] [Usage Questions]
I updated my sqlalchemy to v.1.4.23 in order to take advantage of syntax like:

with Session.begin() as session:
    session.add(object)

However, when attempting to insert some mapped objects, I am getting an
error I do not understand (not for lack of googling!):

site-packages/sqlalchemy/orm/identity.py", line 97, in __contains__
    if key in self._dict:
TypeError: __hash__ method should return an integer

I've tried the insertion both with and without __hash__() defined in the orm class,
makes no difference.  I've tried to set every field in the class in case it was 
blowing up on hashing null cols, makes no difference.
I don't see where this could be coming from.

Here is the class I'm trying to insert an instance of:

```
  class Channel_data(Base):
      __tablename__ = "channel_data"

      net = Column(String(8), info='Network name', primary_key=True, nullable=False)
      sta = Column(String(6), info='Station name', primary_key=True, nullable=False)
      seedchan = Column(String(3), info='Channel name', primary_key=True, nullable=False)
      location = Column(String(2), info='Location code', primary_key=True, nullable=False)
      ondate = Column(DateTime, info='timestamp without time zone', primary_key=True, nullable=False)
      channel = Column(String(8), info='Channel name')
      channelsrc = Column(String(8), info='Channel name')
      inid = Column(Integer, info='Instrument identifier')
      remark = Column(String(60), info='Channel remark')
      unit_signal = Column(Integer, info='Unit lookup key for units of signal response')
      unit_calib  = Column(Integer, info='Unit lookup key for units of calibration input')
      lat = Column(Numeric(asdecimal=False), info='Locations north of the equator have positive latitudes')
      lon = Column(Numeric(asdecimal=False), info='Longitudes are measured positive east of the Greenwich meridian')
      elev = Column(Numeric(asdecimal=False),info='elevation')
      edepth = Column(Numeric(asdecimal=False), info='Emplacement depth relative to elevation')
      azimuth = Column(Numeric(asdecimal=False),info='Azimuth of instrument from North')
      dip = Column(Numeric(asdecimal=False),info='Dip of instrument down from horizontal')
      format_id = Column(Integer, info='Data format lookup key')
      record_length = Column(Integer, info='The exponent (as a power of 2) of the data record length')
      samprate = Column(Numeric(asdecimal=False), info='Data sampling rate [samples/second]')
      clock_drift = Column(Numeric, info='Tolerance value [seconds/sample] used as threshold for time error detection')
      flags = Column(String(27), info='Channel flags')
      offdate = Column(DateTime, info='timestamp without time zone', default=DEFAULT_ENDDATE)
      lddate  = Column(DateTime, info='timestamp without time zone', server_default=text('NOW()'))

      CheckConstraint('azimuth >= 0 and azmiuth <= 360', name='chd01')
      CheckConstraint('clock_drift >= 0.0', name='chd02')
      CheckConstraint('dip >= -90 and dip <= 90', name='chd03')
      CheckConstraint('edepth >= 0.0', name='chd04')
      CheckConstraint('lat >= -90.0 and lat <= 90', name='chd06')
      CheckConstraint('lon >= -180.0 and lon <= 180', name='chd07')
      CheckConstraint('record_length >= 8 and record_length <= 12', name='chd08')

```
I create an object of this class and try to insert but it crashes with the __hash__
error above.

I used the same process with sqlalchemy v0.9 and it worked fine, but now it doesn't.

Any help would be greatly appreciated!

**Comment by mikehagerty** [comment]
Nevermind ... The issue turned out to be a type mismatch between the orm column type (should've been datetime) and the db column type. 

Man, would be great if a *useful* error msg could've been output !!  Is there no checking between the two ?

**Reply by CaselIT** [reply]
strange that it fails like this. can you provide a mvce so we can try improving that error

**Reply by zzzeek** [reply]
has to be that DateTime on the primary key.   but not sure what kind of object it returned that would fail like that.

**Reply by zzzeek** [reply]
would need to know DB driver in use and python version could help here too, as well as of course what that table actually looks like.

**Comment by CaselIT** [comment]
Not related, but note that check constraint defined like that are ignored. you should add them to `__table_args__`: https://docs.sqlalchemy.org/en/14/orm/declarative_tables.html#declarative-table-configuration
